Mapper annotated text plugin highlighter  "array_index_out_of_bounds_exception"

**Describe the feature:**

**Elasticsearch version**:  7.10.0

**Plugins installed**: ["mapper-annotated-text"]

**JVM version**: 
openjdk 11.0.9.1 2020-11-04
OpenJDK Runtime Environment (build 11.0.9.1+1-Ubuntu-0ubuntu1.18.04)
OpenJDK 64-Bit Server VM (build 11.0.9.1+1-Ubuntu-0ubuntu1.18.04, mixed mode, sharing)

**OS version:**
18.04.1-Ubuntu 

**Description of the problem including expected versus actual behavior:**

Mapper annotated text plugin highlighter  "array_index_out_of_bounds_exception".

**Steps to reproduce:**

```
PUT tmp
{
  "mappings": {
      "properties": {
        "my_field": {
          "type": "annotated_text"
        }
      }
    }
}

PUT tmp/_doc/1
{
  "my_field": "[Jeff Beck](Beck) plays a strat"
}

PUT tmp/_doc/2
{
  "my_field": "[Kimchy](Beck) plays a strat"
}

POST tmp/_search
{
  "query": {
    "term": {
        "my_field": "Beck"
    }
  },
  "highlight": {
    "fields": {
      "my_field": {
        "type": "annotated",
        "require_field_match": false
      }
    }
  }
}
```

**Provide logs:**
```
Caused by: org.elasticsearch.ElasticsearchException$1: Index 1 out of bounds for length 1
        at org.elasticsearch.ElasticsearchException.guessRootCauses(ElasticsearchException.java:644) ~[elasticsearch-7.10.0.jar:7.10.0]
        at org.elasticsearch.action.search.AbstractSearchAsyncAction.executeNextPhase(AbstractSearchAsyncAction.java:322) [elasticsearch-7.10.0.jar:7.10.0]
        ... 22 more
Caused by: java.lang.ArrayIndexOutOfBoundsException: Index 1 out of bounds for length 1
        at org.elasticsearch.index.mapper.annotatedtext.AnnotatedTextFieldMapper$AnnotatedHighlighterAnalyzer.lambda$wrapComponents$0(AnnotatedTextFieldMapper.java:320) ~[?:?]
        at org.apache.lucene.analysis.Analyzer$TokenStreamComponents.setReader(Analyzer.java:418) ~[lucene-core-8.7.0.jar:8.7.0 2dc63e901c60cda27ef3b744bc554f1481b3b067 - atrisharma - 2020-10-29 19:35:28]
        at org.apache.lucene.analysis.Analyzer$TokenStreamComponents.access$000(Analyzer.java:366) ~[lucene-core-8.7.0.jar:8.7.0 2dc63e901c60cda27ef3b744bc554f1481b3b067 - atrisharma - 2020-10-29 19:35:28]
        at org.apache.lucene.analysis.Analyzer.tokenStream(Analyzer.java:203) ~[lucene-core-8.7.0.jar:8.7.0 2dc63e901c60cda27ef3b744bc554f1481b3b067 - atrisharma - 2020-10-29 19:35:28]
        at org.apache.lucene.search.uhighlight.AnalysisOffsetStrategy.tokenStream(AnalysisOffsetStrategy.java:56) ~[lucene-highlighter-8.7.0.jar:8.7.0 2dc63e901c60cda27ef3b744bc554f1481b3b067 - atrisharma - 2020-10-29 19:36:06]
        at org.apache.lucene.search.uhighlight.MemoryIndexOffsetStrategy.getOffsetsEnum(MemoryIndexOffsetStrategy.java:94) ~[lucene-highlighter-8.7.0.jar:8.7.0 2dc63e901c60cda27ef3b744bc554f1481b3b067 - atrisharma - 2020-10-29 19:36:06]
        at org.apache.lucene.search.uhighlight.FieldHighlighter.highlightFieldForDoc(FieldHighlighter.java:76) ~[lucene-highlighter-8.7.0.jar:8.7.0 2dc63e901c60cda27ef3b744bc554f1481b3b067 - atrisharma - 2020-10-29 19:36:06]
        at org.apache.lucene.search.uhighlight.CustomFieldHighlighter.highlightFieldForDoc(CustomFieldHighlighter.java:59) ~[elasticsearch-7.10.0.jar:8.7.0 2dc63e901c60cda27ef3b744bc554f1481b3b067 - atrisharma - 2020-10-29 19:36:06]
        at org.apache.lucene.search.uhighlight.CustomUnifiedHighlighter.highlightField(CustomUnifiedHighlighter.java:151) ~[elasticsearch-7.10.0.jar:8.7.0 2dc63e901c60cda27ef3b744bc554f1481b3b067 - atrisharma - 2020-10-29 19:36:06]
        at org.elasticsearch.search.fetch.subphase.highlight.UnifiedHighlighter.highlight(UnifiedHighlighter.java:82) ~[elasticsearch-7.10.0.jar:7.10.0]
        at org.elasticsearch.search.fetch.subphase.highlight.HighlightPhase$1.process(HighlightPhase.java:72) ~[elasticsearch-7.10.0.jar:7.10.0]
        at org.elasticsearch.search.fetch.FetchPhase.execute(FetchPhase.java:167) ~[elasticsearch-7.10.0.jar:7.10.0]
        at org.elasticsearch.search.SearchService.executeFetchPhase(SearchService.java:462) ~[elasticsearch-7.10.0.jar:7.10.0]
        at org.elasticsearch.search.SearchService.executeQueryPhase(SearchService.java:438) ~[elasticsearch-7.10.0.jar:7.10.0]
        at org.elasticsearch.search.SearchService.access$500(SearchService.java:141) ~[elasticsearch-7.10.0.jar:7.10.0]
        at org.elasticsearch.search.SearchService$2.lambda$onResponse$0(SearchService.java:401) ~[elasticsearch-7.10.0.jar:7.10.0]
        at org.elasticsearch.action.ActionRunnable.lambda$supply$0(ActionRunnable.java:58) ~[elasticsearch-7.10.0.jar:7.10.0]
        at org.elasticsearch.action.ActionRunnable$2.doRun(ActionRunnable.java:73) ~[elasticsearch-7.10.0.jar:7.10.0]
        at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37) ~[elasticsearch-7.10.0.jar:7.10.0]
```

--- a/plugins/mapper-annotated-text/src/main/java/org/elasticsearch/index/mapper/annotatedtext/AnnotatedTextFieldMapper.java
+++ b/plugins/mapper-annotated-text/src/main/java/org/elasticsearch/index/mapper/annotatedtext/AnnotatedTextFieldMapper.java
@@ -272,9 +272,16 @@ public class AnnotatedTextFieldMapper extends FieldMapper {
     // The class takes markedup format field values and returns plain text versions.
     // When asked to tokenize plain-text versions by the highlighter it tokenizes the
     // original markup form in order to inject annotations.
+    // Unlike other Analyzers, which tend to be single-instance, this class has
+    // instances created per search request and field being highlighted. This allows us to keep
+    // state about the annotations being processed and pass them into token streams
+    // being highlighted.
     public static final class AnnotatedHighlighterAnalyzer extends AnalyzerWrapper {
         private final Analyzer delegate;
         private AnnotatedText[] annotations;
+        // If the field has arrays of values this counter is used to keep track of
+        // which array element is currently being highlighted.
+        AtomicInteger readerNum;
 
         public AnnotatedHighlighterAnalyzer(Analyzer delegate){
             super(delegate.getReuseStrategy());
@@ -286,14 +293,15 @@ public class AnnotatedTextFieldMapper extends FieldMapper {
           return delegate;
         }
 
+        // Called with each new doc being highlighted
         public void setAnnotations(AnnotatedText[] annotations) {
-            this.annotations = annotations;
+            this.annotations = annotations;            
+            this.readerNum = new AtomicInteger(0);
         }
 
         @Override
         protected TokenStreamComponents wrapComponents(String fieldName, TokenStreamComponents components) {
             AnnotationsInjector injector = new AnnotationsInjector(components.getTokenStream());
-            AtomicInteger readerNum = new AtomicInteger(0);
             return new TokenStreamComponents(r -> {
                 String plainText = readToString(r);
                 AnnotatedText at = annotations[readerNum.getAndIncrement()];

--- a/plugins/mapper-annotated-text/src/main/java/org/elasticsearch/index/mapper/annotatedtext/AnnotatedTextFieldMapper.java
+++ b/plugins/mapper-annotated-text/src/main/java/org/elasticsearch/index/mapper/annotatedtext/AnnotatedTextFieldMapper.java
@@ -266,15 +266,18 @@ public class AnnotatedTextFieldMapper extends FieldMapper {
         }
     }
 
-    // A utility class for use with highlighters where the content being highlighted
-    // needs plain text format for highlighting but marked-up format for token discovery.
-    // The class takes markedup format field values and returns plain text versions.
-    // When asked to tokenize plain-text versions by the highlighter it tokenizes the
-    // original markup form in order to inject annotations.
-    // Unlike other Analyzers, which tend to be single-instance, this class has
-    // instances created per search request and field being highlighted. This allows us to keep
-    // state about the annotations being processed and pass them into token streams
-    // being highlighted.
+    /**
+     * A utility class for use with highlighters where the content being highlighted
+     * needs plain text format for highlighting but marked-up format for token discovery.
+     * The class takes marked up format field values and returns plain text versions.
+     * When asked to tokenize plain-text versions by the highlighter it tokenizes the
+     * original markup form in order to inject annotations.
+     * WARNING - not thread safe.
+     * Unlike other Analyzers, which tend to be single-instance, this class has
+     * instances created per search request and field being highlighted. This allows us to 
+     * keep state about the annotations being processed and pass them into token streams
+     * being highlighted.
+     */
     public static final class AnnotatedHighlighterAnalyzer extends AnalyzerWrapper {
         private final Analyzer delegate;
         private AnnotatedText[] annotations;

--- a/plugins/mapper-annotated-text/src/main/java/org/elasticsearch/index/mapper/annotatedtext/AnnotatedTextFieldMapper.java
+++ b/plugins/mapper-annotated-text/src/main/java/org/elasticsearch/index/mapper/annotatedtext/AnnotatedTextFieldMapper.java
@@ -281,7 +281,7 @@ public class AnnotatedTextFieldMapper extends FieldMapper {
         private AnnotatedText[] annotations;
         // If the field has arrays of values this counter is used to keep track of
         // which array element is currently being highlighted.
-        AtomicInteger readerNum;
+        int readerNum;
 
         public AnnotatedHighlighterAnalyzer(Analyzer delegate){
             super(delegate.getReuseStrategy());
@@ -296,7 +296,7 @@ public class AnnotatedTextFieldMapper extends FieldMapper {
         // Called with each new doc being highlighted
         public void setAnnotations(AnnotatedText[] annotations) {
             this.annotations = annotations;            
-            this.readerNum = new AtomicInteger(0);
+            this.readerNum = 0;
         }
 
         @Override
@@ -304,7 +304,7 @@ public class AnnotatedTextFieldMapper extends FieldMapper {
             AnnotationsInjector injector = new AnnotationsInjector(components.getTokenStream());
             return new TokenStreamComponents(r -> {
                 String plainText = readToString(r);
-                AnnotatedText at = annotations[readerNum.getAndIncrement()];
+                AnnotatedText at = annotations[readerNum++];
                 assert at.textMinusMarkup.equals(plainText);
                 injector.setAnnotations(at);
                 components.getSource().accept(new StringReader(at.textMinusMarkup));

--- a/plugins/mapper-annotated-text/src/main/java/org/elasticsearch/index/mapper/annotatedtext/AnnotatedTextFieldMapper.java
+++ b/plugins/mapper-annotated-text/src/main/java/org/elasticsearch/index/mapper/annotatedtext/AnnotatedTextFieldMapper.java
@@ -54,7 +54,6 @@ import java.util.Arrays;
 import java.util.List;
 import java.util.Map;
 import java.util.Objects;
-import java.util.concurrent.atomic.AtomicInteger;
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
 

--- a/plugins/mapper-annotated-text/src/yamlRestTest/resources/rest-api-spec/test/mapper_annotatedtext/10_basic.yml
+++ b/plugins/mapper-annotated-text/src/yamlRestTest/resources/rest-api-spec/test/mapper_annotatedtext/10_basic.yml
@@ -144,6 +144,36 @@
   - match: {_shards.failed: 0}
 
 ---
+"issue 66535 array index out of bounds issue -requires multiple docs":
+  - do:
+      indices.create:
+        index: annotated
+        body:
+          mappings:
+            properties:
+              my_field:
+                type: annotated_text
+
+  - do:
+      index:
+        index: annotated
+        id: 1
+        body:
+            "my_field" : "[Jeff Beck](Beck) plays a strat"
+  - do:
+      index:
+        index: annotated
+        id: 2
+        body:
+            "my_field" : "[Kimchy](Beck) plays a strat"
+        refresh: true
+  - do:
+      search:
+        request_cache: false
+        body: {"query": {"term": {"my_field": "Beck"}  }, "highlight": {"fields": {"my_field": {"type": "annotated", "require_field_match": false }}}}
+  - match: {_shards.failed: 0}
+
+---
 "Signficant text support":
   - do:
       indices.create:

